sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/auth/CachedTokenProvider.java

```diff
@@ -0,0 +1,68 @@
+package com.mesosphere.sdk.dcos.auth;
+
+import com.auth0.jwt.interfaces.DecodedJWT;
+
+import java.io.IOException;
+import java.time.Instant;
+import java.util.Optional;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/**
+ * CachedTokenProvider retrieves token from underlying provider and caches the value. It automatically triggers
+ * getToken() method on underlying provider when token is about to expire.
+ */
+public class CachedTokenProvider implements TokenProvider {
```

Comment:
Ah, didn't realize we weren't using it yet.

I'd stick with what you have in that case.